dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/WebSphereCDIDeploymentImpl.java

```diff
@@ -586,15 +602,54 @@ public WebSphereBeanDeploymentArchive getBeanDeploymentArchiveFromClass(Class<?>
                     }
                     if (bdaClazz == clazz) {
 
-                        wbda = bda;
-                        classBDAMap.put(clazz, bda);
-                        break;
+                        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                            Tr.debug(tc, "mapping class: " + clazz.getName() + " " + System.identityHashCode(clazz) + " to bda: " + bda.toString());
+                        }
+
+                        //Always populate classBDAMap with the first BDA that contains this class
+                        classBDAMap.putIfAbsent(clazz, bda);
+
+                        //Do a quick check to see if this is also the first EJB_MODULE that contains the class and if so record it.
+                        if (bda.getType() == ArchiveType.EJB_MODULE) {
+                            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                                Tr.debug(tc, "mapping class: " + clazz.getName() + " " + System.identityHashCode(clazz) + " to EJB_MODULE bda: " + bda.toString());
+                            }
+                            classBDAMapFavouringEjb.putIfAbsent(clazz, bda);
+                        }
+
+                        //If we are not in favour EJB mode, check the standard map, if we've put something there return that.
+                        //If we are in favour EJB mode
+                        if (!favourEJB && classBDAMap.containsKey(clazz)) {
+                            wbda = classBDAMap.get(clazz);
+                            break;
+                        } else if (classBDAMapFavouringEjb.containsKey(clazz)) {
+                            wbda = classBDAMapFavouringEjb.get(clazz);
+                            break;
+                        }
                     }
                 }
             }
         }
 
-        return wbda;
+        //If we've reached here in EJB mode, cache that we don't have a value
+        if (favourEJB) {
+            classBDAMapFavouringEjb.putIfAbsent(clazz, new MapNoValueMarker());
+        }
+
+        //Always return the standard map's value as a fallback
+        return classBDAMap.get(clazz);
+    }
+
+    @Trivial
+    @Override
+    public WebSphereBeanDeploymentArchive getBeanDeploymentArchiveFromClassFavouringEJB(Class<?> clazz) {
+
+        if (classBDAMapFavouringEjb.containsKey(clazz)) {
+            return classBDAMapFavouringEjb.get(clazz);
+        }
+
+        //If we haven't cached the value, call the internal. It will populate the map, and handle falling back to any BDA with the right class
+        return getBeanDeploymentArchiveFromClassInternal(clazz, true);
```

Comment:
As a general note, I really don't like that we're checking the cache map in one method but writing it in another.

Here I think it's actually wrong, we'd return the `MapNoValueMarker` object here and `CDIEJBManagedObjectFactory` isn't expecting that.